build.rs

```diff
@@ -218,6 +387,167 @@ fn build_msvc(target_triple: &[&str], disable_opt: bool, num_jobs: &str,
+}
+
+fn build_unix(target_info: &TargetInfo, out_dir: PathBuf) {
+    let mut lib_target = out_dir.clone();
+    lib_target.push("libring-core.a");
+    let lib_target = lib_target.as_path();
+
+    let mut test_target = out_dir.clone();
+    test_target.push("libring-test.a");
+    let test_target = test_target.as_path();
+
+    let lib_header_change = RING_HEADERS.par_iter()
+        .chain(RING_INLINE_FILES.par_iter())
+        .chain(RING_BUILD_FILE.par_iter())
+        .map(Path::new)
+        .any(|p| need_run(&p, lib_target));
+    let test_header_change = RING_TEST_HEADERS.par_iter()
+        .map(Path::new)
+        .any(|p| need_run(&p, test_target)) ||
+                             lib_header_change;
+
+    let srcs = match target_info.target_arch() {
+        "x86_64" => vec![RING_X86_64_SRC, RING_INTEL_SHARED_SRCS],
+        "x86" => vec![RING_X86_SRCS, RING_INTEL_SHARED_SRCS],
+        "arm" => vec![RING_ARM_SHARED_SRCS, RING_ARM_SRCS],
+        "aarch64" => vec![RING_ARM_SHARED_SRCS, RING_AARCH64_SRCS],
+        _ => Vec::new(),
+    };
+
+    let additional = srcs.into_par_iter()
+        .weight_max()
+        .flat_map(|additional_src| {
+            additional_src.par_iter()
+                .map(|src| make_asm(src, out_dir.clone(), &target_info))
+        });
+    build_library(lib_target,
+                  additional,
+                  RING_SRC,
+                  target_info,
+                  out_dir.clone(),
+                  lib_header_change);
+
+    // XXX: Ideally, this would only happen for `cargo test`,
+    // but we don't know how to do that yet.
+    build_library(test_target,
+                  Vec::new().into_par_iter(),
+                  RING_TEST_SRCS,
+                  target_info,
+                  out_dir.clone(),
+                  test_header_change);
+
+    // target_vendor is only set if a nightly version of rustc is used
+    let libcxx = if target_info.target_vendor()
+        .map(|v| v == "apple")
+        .unwrap_or(target_info.target_os() == "macos" ||
+                   target_info.target_os() == "ios") ||
+                    target_info.target_os() == "freebsd" {
+        "c++"
+    } else {
+        "stdc++"
+    };
+    println!("cargo:rustc-flags=-l dylib={}", libcxx);
+    print_rerun();
+}
+
+
+fn build_library<P>(target: &Path, additional: P,
+                    lib_src: &'static [&'static str],
+                    target_info: &TargetInfo, out_dir: PathBuf,
+                    header_changed: bool)
+    where P: ParallelIterator<Item = String>
+{
+    let objs = additional.chain(lib_src.par_iter().map(|a| String::from(*a)))
+        .weight_max()
+        .map(|f| compile(&f, &target_info, out_dir.clone(), header_changed))
+        .map(|v| vec![v])
+        .reduce(Vec::new,
+                &|mut a: Vec<String>, b: Vec<String>| -> Vec<String> {
+                    a.extend(b.into_iter());
+                    a
+                });
+    if objs.par_iter()
```

Comment:
Please add a comment here about what this does. I think it would be something like "Rebuild the library if necessary."